pkg/crd/flatten.go

```diff
@@ -26,47 +27,176 @@ import (
+	case reflect.Chan, reflect.Func, reflect.Interface, reflect.Map, reflect.Ptr, reflect.Slice:
+		return val.IsNil()
+	default:
+		return valInt == zeroInt
+	}
+}
+
 // flattenAllOfInto copies properties from src to dst, then copies the properties
 // of each item in src's allOf to dst's properties as well.
-func flattenAllOfInto(dst *apiext.JSONSchemaProps, src apiext.JSONSchemaProps) {
+func flattenAllOfInto(dst *apiext.JSONSchemaProps, src apiext.JSONSchemaProps, errRec ErrorRecorder) {
 	if len(src.AllOf) > 0 {
 		for _, embedded := range src.AllOf {
-			flattenAllOfInto(dst, embedded)
+			flattenAllOfInto(dst, embedded, errRec)
 		}
 	}
 
-	for propName, prop := range src.Properties {
-		dst.Properties[propName] = prop
+	dstVal := reflect.Indirect(reflect.ValueOf(dst))
+	srcVal := reflect.ValueOf(src)
+	typ := dstVal.Type()
+
+	srcRemainder := apiext.JSONSchemaProps{}
+	srcRemVal := reflect.Indirect(reflect.ValueOf(&srcRemainder))
+	dstRemainder := apiext.JSONSchemaProps{}
+	dstRemVal := reflect.Indirect(reflect.ValueOf(&dstRemainder))
+	hoisted := false
+
+	for i := 0; i < srcVal.NumField(); i++ {
+		fieldName := typ.Field(i).Name
+		if fieldName == "AllOf" {
+			continue
+		}
+		srcField := srcVal.Field(i)
+		fldTyp := srcField.Type()
+		zeroVal := reflect.Zero(fldTyp)
+		zeroInt := zeroVal.Interface()
+		srcInt := srcField.Interface()
+
+		if isOrNil(srcField, srcInt, zeroInt) {
+			// nothing to copy from src, continue
+			continue
+		}
+
+		dstField := dstVal.Field(i)
+		dstInt := dstField.Interface()
+		if isOrNil(dstField, dstInt, zeroInt) {
+			// dst is empty, continue
+			dstField.Set(srcField)
+			continue
+		}
+
+		if fldTyp.Comparable() && srcInt == dstInt {
+			// same value, continue
+			continue
+		}
+
+		// resolve conflict
+		switch fieldName {
+		case "Properties":
+			// merge if possible, use all of otherwise
+			srcMap := srcInt.(map[string]apiext.JSONSchemaProps)
+			dstMap := dstInt.(map[string]apiext.JSONSchemaProps)
+
+			for k, v := range srcMap {
+				dstProp, exists := dstMap[k]
+				if !exists {
+					dstMap[k] = v
+					continue
+				}
+				flattenAllOfInto(&dstProp, v, errRec)
+				dstMap[k] = dstProp
+			}
+		case "Required":
+			// merge
+			dstField.Set(reflect.AppendSlice(dstField, srcField))
+		case "Type":
+			if srcInt != dstInt {
+				// TODO(directxman12): figure out how to attach this back to a useful point in the Go source or in the schema
+				errRec.AddError(fmt.Errorf("conflicting types in allOf branches in schema: %s vs %s", dstInt, srcInt))
+			}
+			// keep the destination value, for now
+		// TODO(directxman12): Default -- use field?
+		// TODO(directxman12):
+		// - Dependencies: if field x is present, then either schema validates or all props are present
+		// - AdditionalItems: like AdditionalProperties
+		// - Definitions: common named validation sets that can be references (merge, bail if duplicate)
+		case "AdditionalProperties":
+			// TODO(directxman12): handle the bool too
```

Comment:
we don't support `false` if I remember right. `additionalSchema: true` is a no-op logically, but has influence on pruning.

Reply:
that's what I thought, but wanted to check with you first.  good to confirm.